README.md
# 全球性 流感 传播过程 重构 平台

Reconfigurable Platform of Spreading Process of Global Influenza.

## screenshot

初始页面：

![](./docs/screenshot1.png)

重构演示：

![](./docs/screenshot2.png)

降维结果展示：

![](./docs/screenshot3.png)

## User Guide

[Guide.md](./docs/Guide.md)

## Functions

该平台基于Python3.6语言开发, 主要实现了以下两方面功能：

- 以人工导入的方式实现NCBI数据库 (https://www.ncbi.nlm.nih.gov/genomes/FLU/Database/nph-select.cgi?go=database) 中流感疫情监控数据的导入，包括流感基因氨基酸序列和全球流感发生地城市的经纬度信息，并自动提取每一流感样本的爆发时间、地点等信息，再按照时间顺序将所有病毒样本及其对应的爆发时间、地点等信息在世界地图中重现，用以重构病毒的全球性传播过程。

- 该平台还可以自动提取出氨基酸序列中不同流感对应的氨基酸序列信息，在PCA降维处理之后，进一步进行可视化表示，以一种更为直观的方式来发现流感病毒的进化趋势。


This platform is designed on the base of python3.6 language. Its operating process can be divided into following steps: 

- For the first step, manually inputting monitoring data of influenza from NCBI(National Center for Biotechnology Information), including amino acid sequence file and the global urban latitude and longitude information file. For the second step, the program automatically extracts information about the time and place of the influenza outbreak in the file. For the last step, the program reconstructs all the flu information and its corresponding outbreak time and location in the world map in chronological order. 

- This program can automatically extract the amino acid sequence of different influenza and corresponding sequence information, and then the amino acid sequence will be processed on PCA data dimension reduction, and finally the processing results are clearly displayed on the reconfigurable platform. In this way, the evolving trend of influenza virus can be obtained directly. 


## Any advice or sugggestions

Any advice or sugggestions, please mail to b4zinga@outlook.com

docs/CHANGELOG.md
## 2018-08-06

1. 添加至 github 


## 2018-4-21


1. 在入口fluplatform.py中添加 version check


## 2018-4-19


1. 将预处理功能的弹出多个窗口修改为个数超过默认值后(setting.py中的UnMatchCityNum, 默认为10)，将城市信息输出至文件(./LostLocationCities.txt)中，提示用户手动添加


operfile.py
#!/usr/bin/env python
# coding: utf-8
# Date  : 2018-08-06 11:40:23
# Author: b4zinga
# Email : b4zinga@outlook.com
# Func  : operate files

import re
from collections import Counter


class FA:
    """FA file is the protein FASTA file download for NCBI.
    BCBI(https://www.ncbi.nlm.nih.gov/genomes/FLU/Database/nph-select.cgi#mainform).
    """
    def __init__(self, fa):
        """Initialization.
        time_format_err is the record of false time formate in FA file.

        """
        self.fa = fa
        self.time_format_err = []

    def getPlaceAndTime(self):
        """

        return a list like that: [('South Carolina', '2012/10/04'),]
        """
        place_time = []

        # sample:
        # >AEK85498 A/Taiwan/90252/2011 2011/02/ HA
        # >BAI48182 A/Nagasaki/HA-29/2009 2009/07/30 HA
        # >ABK40557 A/Memphis/27/1983 1983// HA
        regex = ">.*?\s\w/(.*?)/.*?/\d+\s(\d+/\d+/\d+)\s.*?$"  # low
        # regex = '>.*?\s\w/(.*?)/.*?/\d+\s+(\d+/\d*/\d*)\s+HA'  # middle
        # regex = '>.*?\s\w/(.*?)/\d+\s+(\d+/\d*/\d*)\s.*?$'     # high

        with open(self.fa) as file:
            for f in file:
                if f.startswith('>'):
                    items = re.findall(regex, f)
                    if items:
                        place_time.append(items[0])
                    else:
                        self.time_format_err.append(f)

        return place_time

    def getDiffPlaceTime(self):
        """get the amount of flu erupting in the same place and time.
        return a dict like that: {('New Jersey', '1976//'): 3, ('HaNoi', '2003/01/26'): 1,}
        """
        return dict(Counter(self.getPlaceAndTime()))

    def getTimeFormatErr(slef):
        return self.time_format_err

    def getVirusAmino(self):
        """
        """
        pass


class City:
    """City file involves the list of global cities relating to longitude and latitude.
    """
    def __init__(self, city):
        """Initialization.
        city_format_err is the record of false format in city file.
        """
        self.city = city
        self.city_format_err = []

    def getCityLocation(self):
        """get all the city names and corresponding longitude and latitude in city file.
        return a dict like that: {'Tientsin': ('39.084158', '117.200983'),}
        """
        city_location = {}
        # sample:
        #  58.001985     56.257287      Perm
        #  53.6126505    12.4295953     Germany-MV(Mecklenburg-Vorpommern)
        regex = ".?([-]?\d+[\.]?\d+)\s+([-]?\d+[\.]?\d+)\s+(.*?)[(\n]"
        with open(self.city) as file:
            for f in file:
                items = re.findall(regex, f)
                if items:
                    for item in items:
                        city_location[item[2]] = (item[:2])
                else:
                    self.city_format_err.append(f)

        return city_location

    def getCityFormatErr(self):
        """return a list of false format of longitude and latitude.
        """
        return self.city_format_err

    def addCityIntoFile(self, jing, wei, place):
        """add longitude and latitude to city file.
        """
        info = ' ' + wei + '    ' +jing + '    ' + place + '\n'
        with open(self.city, 'a') as cf:
            cf.write(info)
        cf.close()


class Operation(FA, City):
    """Some methods based on the FA file and City file."""
    def __init__(self, fa, city):
        """Initialization.
        match_failed_record is records that city names can be found in FA file, 
        but corresponding longitude and latitude can not be found in city file.
        """
        super().__init__(fa)
        super(FA, self).__init__(city)

        self.match_failed_record = []

    def getExplosionDetail(self):
        """return a list like that: [['Singapore', '2009/09/23', ('1.352083', '103.819836'), 3],]
        """
        details = []
        place_and_time = self.getDiffPlaceTime()
        place_jing_wei = self.getCityLocation()
        for pt, num in place_and_time.items():
            try:
                details.append([pt[0], pt[1], place_jing_wei[pt[0]], num])
            except KeyError:
                self.match_failed_record.append((pt,num),)

        return sorted(details, key=lambda item: item[1])

    def getMatchFailedRecord(self):
        """match_failed_record is records that city names can be found in FA file, 
        but corresponding longitude and latitude can not be found in city file.
        """
        return self.match_failed_record

    def judgeIntegrity(self):
        """judging the intergrity of city file, which means whether city names in
        FA file can be matched with corresponding longitude and latitude in city
        file or not.
        If matched, return False, else return unmatched city names.
        """
        place_and_time = self.getPlaceAndTime()
        city_location = self.getCityLocation()
        no_exist_place = set()
        for pt in place_and_time:
            if pt[0] not in city_location.keys():
                no_exist_place.add(pt[0])
        if len(no_exist_place):
            return no_exist_place
        else:
            return False

    def getDetailByTime(self):
        """get corresponding information according to the order of flu erruption.
        return a dict like that : 
        {'2010/09/16': [['Sydney', '2010/09/16', ('-33.8674869', '151.2069902'), 2], ['Bangkok', '2010/09/16', ('13.7522222', '100.4938889'), 1]], }
        and the start time and end time of influenza.
        """
        details_by_time = {}
        virus = self.getExplosionDetail()
        start_time, end_time = virus[0][1], virus[-1][1]
        for vir in virus:
            details_by_time[vir[1]] = []
        for vir in virus:
            details_by_time[vir[1]].append(vir)

        return details_by_time, start_time, end_time



if __name__ == '__main__':
    op = Operation('./data/H1N1_1965-2012.fa', './data/LatLon_H1N1.txt')
    print(op.getDetailByTime())


clustering.py
#!/usr/bin/env python
# coding: utf-8
# Date  : 2018-08-06 11:38:56
# Author: b4zinga
# Email : b4zinga@outlook.com
# Func  :

import re
import numpy as np
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
# from pca import PCA

from settings import AMINO


class FrequencyCluster:
    """Calculate the frequency of amino acid sequences of
    virus in the FA file, and Cluster it with PCA
    """
    def __init__(self, fa):
        self.fa = fa

    def loadDate(self):
        """Take out all amino acid sequences of each virus in the FA file.
        return a dict like that: {('ARQ87434', '2017/04/06'):'MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVL...',}
        """
        amino_acid_sequences = {}
        regex = '>(.*?)\s\w/.*?/\d+\s+(\d+/\d*/\d*)\s.*?$'
        with open(self.fa) as file:
            for f in file:
                if f.startswith('>'):
                    items = re.findall(regex, f)
                    if items:
                        virus_name = items[0]
                        amino_acid_sequences[virus_name] = ''
                    else:
                        virus_name = None
                else:
                    try:
                        amino_acid_sequences[virus_name] += f.strip('\n')
                    except KeyError:
                        pass

        return amino_acid_sequences

    def calcWeight(self):
        """Calc of the frequency of amino in acid sequence.
        return a list like that: [[34, 16, 23, 33, 21, 44, 14, 35, 37, 48, 8, 45, 21, 16, 20, 43, 38, 34, 10, 26],...]
        """
        virus = self.loadDate()  # {, 'ARQ87217': 'MKAILVVLLYTFTTANADTLC...',...}
        amino_frequency = []
        for vir in virus:
            amino = AMINO
            temp = []
            for am in amino:
                temp.append(virus[vir].count(am))
            amino_frequency.append(temp)

        return amino_frequency

    def cluster(self):
        """cluster the frequency data with PCA.
        return the <class 'numpy.matrixlib.defmatrix.matrix'> after PCA:

        [[ 2.73957257  4.37035856]
         [ 4.89129411  3.42768334]
         [ 3.98168808  2.79871617]
         ...
         [-2.05245831  4.4499471 ]
         [-2.63118146  4.34273584]
         [-2.40254118  5.18975161]]
        """
        list_data = self.calcWeight()
        weight = np.array(list_data, dtype=int)
        pca = PCA(n_components=2)
        new_data = pca.fit_transform(weight)
        data_mat = np.mat(new_data)

        return data_mat

    def show(self, marker='.', s=1, color='RdYlBu_r'):
        """cluster the frequency data with PCA and show the result.

        list_data: the frequency list of amino acid sequence.
        marker   : scatter markers , the marker style.
        s        : scatter s , the size of markers.
        color    : the color show in PCA.

        Possible values are: Accent, Accent_r, Blues, Blues_r,
        BrBG, BrBG_r, BuGn, BuGn_r, BuPu, BuPu_r, CMRmap, CMRmap_r,
        Dark2, Dark2_r, GnBu, GnBu_r, Greens, Greens_r, Greys,
        Greys_r, OrRd, OrRd_r, Oranges, Oranges_r, PRGn, PRGn_r,
        Paired, Paired_r, Pastel1, Pastel1_r, Pastel2, Pastel2_r,
        PiYG, PiYG_r, PuBu, PuBuGn, PuBuGn_r, PuBu_r, PuOr, PuOr_r,
        PuRd, PuRd_r, Purples, Purples_r, RdBu, RdBu_r, RdGy,
        RdGy_r, RdPu, RdPu_r, RdYlBu, RdYlBu_r, RdYlGn, RdYlGn_r,
        Reds, Reds_r, Set1, Set1_r, Set2, Set2_r, Set3, Set3_r,
        Spectral, Spectral_r, Wistia, Wistia_r, YlGn, YlGnBu,
        YlGnBu_r, YlGn_r, YlOrBr, YlOrBr_r, YlOrRd, YlOrRd_r ...
        其中末尾加r是颜色取反
        """
        list_data = self.calcWeight()
        weight = np.array(list_data, dtype=int)
        pca = PCA(n_components=2)
        new_data = pca.fit_transform(weight)
        data_mat = np.mat(new_data)

        fig = plt.figure()
        ax = fig.add_subplot(111)
        cm = plt.cm.get_cmap(color)
        sc = ax.scatter(x=data_mat[:, 0].flatten().A[0],
                        y=data_mat[:, 1].flatten().A[0],
                        c=data_mat[:, 1].flatten().A[0]*0.2 + data_mat[:, 0].flatten().A[0]*0.8,
                        # c=data_mat[:, 1].flatten().A[0],
                        # c=data_mat[:, 0].flatten().A[0],
                        marker=marker,
                        s=s,
                        vmin=0, vmax=20,
                        cmap=cm)
        plt.grid(ls=':')
        plt.colorbar(sc)
        plt.show()        



if __name__ == '__main__':
    fc = FrequencyCluster('data/H1N1_1976-2017_without_XXX.fa')
    fc.show()

showdata.py
#!/usr/bin/env python
# coding: utf-8
# Date  : 2018-08-06 11:41:30
# Author: b4zinga
# Email : b4zinga@outlook.com
# Func  :

import os
import time
import base64
import datetime
from dateutil.relativedelta import relativedelta

from tkinter import *
from tkinter import messagebox
from tkinter.filedialog import askopenfilename

from operfile import Operation
from clustering import FrequencyCluster
from settings import ABOUT, ICO, BGMAP, DailyLength, VirusLiveTime, UnMatchCityNum, LostLocationCities, HelpInfo


RUNNING = False  # global variable, whether pause or not.
INTERVAL = datetime.timedelta(days=1)  # global variable, time interval,  default :1 day


class Virus:
    """Define the basic information of influenza.
    """
    def __init__(self):
        self.VIRUS_LIVE_TIME = IntVar()  # Virus survival time (day)
        self.BASE_DAY_TIME = IntVar()  # Daily corresponding time (seconds)
        self.set_virus_live_time()  
        self.set_base_day_time()

    def set_virus_live_time(self, day=VirusLiveTime):
        self.VIRUS_LIVE_TIME.set(day)

    def set_base_day_time(self, seconds=DailyLength):
        self.BASE_DAY_TIME.set(seconds)

    def get_virus_live_time(self):
        return self.VIRUS_LIVE_TIME.get()

    def get_base_day_time(self):
        return self.BASE_DAY_TIME.get()

    def get_virus_live_time_obj(self):
        return self.VIRUS_LIVE_TIME

    def get_base_day_time_obj(self):
        return self.BASE_DAY_TIME


# TODO: use base64 encode bytes to replace temporary .png file.
class Application(Tk):
    """Basic Map Application
    """
    def __init__(self):
        """Initialization
        """
        super().__init__()
        self.createBg()
        self.createIcon()
        self.map = PhotoImage(file='map.png')
        self.can = self.createMap()

    @staticmethod
    def createBg():
        """Create 'map.png' file 
        """
        map_tmp = open("map.png", "wb+")
        map_tmp.write(base64.b64decode(BGMAP))
        map_tmp.close()

    @staticmethod
    def createIcon():
        """Create 'ico.ico' file
        """
        ico_tmp = open('ico.ico', 'wb+')
        ico_tmp.write(base64.b64decode(ICO))
        ico_tmp.close()

    def createMap(self):
        """Create Map
        """
        self.title('地图程序')

        cv = Canvas(self, width=800, height=496, bg='SkyBlue')
        cv.create_image(400, 248, image=self.map)
        os.remove("map.png")
        os.remove("ico.ico")
        cv.grid(row=0, column=0)
        return cv

    def addFunc(self, label_frame):
        label_frame(self, self.can)



class FunctionMenu:
    def __init__(self, root, canvas):
        """Initialization
        """
        self.root = root
        self.root.resizable(False, False)
        # Close button on the upper right corner
        # self.root.protocol("WM_DELETE_WINDOW",self.ask_quit)  
        self.canvas = canvas
        self.virus = Virus()
        # Definition of some variables
        self.fa_path = StringVar()
        self.city_path = StringVar()
        self.show_city = IntVar()
        self.show_virus_num = IntVar()
        self.evolution_speed = IntVar()
        self.from_now = StringVar()
        self.from_now.set('如：2017-01-01')

        self.createFuncFrame()

    def createFuncFrame(self):
        """Function GUI"""
        menu_bar = Menu(self.root)
        menu_bar.add_command(label='修改默认配置', command=self.confVirus)
        menu_bar.add_command(label='关于', command=self.about)
        self.root.config(menu=menu_bar)  # == self.root['menu'] = menu_bar

        frm_main = LabelFrame(self.root, text='功能')
        frm_main.grid(row=0, column=1)

        frm_info = Frame(frm_main)
        frm_info.grid(row=0, column=1)
        Label(frm_info, text='流感存活时间 ').grid(row=1, column=1)
        Label(frm_info, textvariable=self.virus.get_virus_live_time_obj()).grid(row=1, column=2)
        Label(frm_info, text='天, 每天显示').grid(row=1, column=3)
        Label(frm_info, textvariable=self.virus.get_base_day_time_obj()).grid(row=1, column=4)
        Label(frm_info, text='秒').grid(row=1, column=5)
        Label(frm_info, text='').grid(row=2, column=1)

        frm_file = Frame(frm_main)
        frm_file.grid(row=1, column=1)
        Entry(frm_file, textvariable=self.fa_path, ).grid(row=0, column=0)
        Button(frm_file, text="选择fa文件", command=self.selectFaFile, cursor="hand2").grid(row=0, column=1)
        Entry(frm_file, textvariable=self.city_path, ).grid(row=1, column=0, pady=5)
        Button(frm_file, text="选择city文件", command=self.selectCityFile, cursor="hand2").grid(row=1, column=1, pady=5)

        frm_ctrl = Frame(frm_main)
        frm_ctrl.grid(row=2, column=1)
        Checkbutton(frm_ctrl, text='显示城市名', variable=self.show_city, cursor="hand2").grid(row=0, column=0)
        Checkbutton(frm_ctrl, text='显示流感数量', variable=self.show_virus_num, cursor="hand2").grid(row=0, column=1)

        frm_speed = Frame(frm_main, padx=10)  # -------------the frame of change speed button
        frm_speed.grid(row=3, column=1)
        Button(frm_speed, text=" + ", command=self.faster, cursor="hand2").grid(row=1, column=0)
        Label(frm_speed, text="< 速 度 >").grid(row=1, column=1)
        Button(frm_speed, text=" - ", command=self.slower, cursor="hand2").grid(row=1, column=2)

        frm_jump = Frame(frm_main, padx=10, pady=10)
        frm_jump.grid(row=4, column=1)
        Button(frm_jump, text='<<上年', command=self.lastYear, cursor="hand2").grid(row=0, column=0)
        Button(frm_jump, text='<上月', command=self.lastMonth, cursor="hand2").grid(row=0, column=1)
        Button(frm_jump, text='暂停', command=self.pause).grid(row=0, column=2)
        Button(frm_jump, text='下月>', command=self.nextMonth, cursor="hand2").grid(row=0, column=3)
        Button(frm_jump, text='下年>>', command=self.nextYear, cursor="hand2").grid(row=0, column=4)

        frm_time = Frame(frm_main)
        frm_time.grid(row=5, column=1)
        Entry(frm_time, width=15, textvariable=self.from_now).grid(row=0, column=0)
        self.btn_fn = Button(frm_time, text='从当前时间开始显示', command=self.btnFromNow, cursor="hand2")
        self.btn_fn.grid(row=0, column=1)

        frame_show = Frame(frm_main, padx=10, pady=10)
        frame_show.grid(row=6, column=1)
        self.btn_p = Button(frame_show, text='预处理', command=self.btnPreprocessing, cursor="hand2")
        self.btn_p.grid(row=0, column=0)
        self.btn_s = Button(frame_show, text='开始画图', bg='PeachPuff', command=self.btnStart, cursor="hand2")
        self.btn_s.grid(row=0, column=1)
        Button(frame_show, text='清除全部', bg='PeachPuff', command=self.clearAll, cursor="hand2").grid(row=0, column=2)
        Button(frame_show, text='退  出', command=self.askQuit).grid(row=0, column=3)

        frm_clustering = Frame(frm_main, padx=10)
        frm_clustering.grid(row=7, column=1)
        Button(frm_clustering, text="显示聚类图", command=self.showClustering, width=28).grid(row=0, column=1)

    @staticmethod
    def convertTime(t):
        """Converts the date format into the format like that: 2016/01/02
        """
        t = t.replace('-', '/').replace('.', '/').replace(' ', '/')
        try:
            result = datetime.datetime.strptime(t, "%Y/%m/%d")
        except ValueError:
            messagebox.showwarning('Warning', 'Time Format Error')
            return
        return result

    @staticmethod
    def convertJingWei(j, w):
        """convert the real Latitude and longitude into the location on map
        """
        j = 407 + (378 / 180) * float(j)
        w = 248 - (218 / 90) * float(w)
        return j, w


    def confVirus(self):
        """To configure the live time of influenza and daily length.
        """
        def btnOk():
            self.virus.set_virus_live_time(vlt.get())
            self.virus.set_base_day_time(bdt.get())
            conf.destroy()

        conf = Toplevel(self.root)
        conf.title('配置信息')
        conf.resizable(False, False)
        conf.attributes("-topmost", 1)  # WinSetOnTop. or conf.wm_attributes("-topmost", 1)
        Label(conf, text='\t当前存活时间：').grid(row=1, column=1, pady=15)
        vlt = Entry(conf, textvariable=self.virus.get_virus_live_time_obj(), width=10)
        vlt.grid(row=1, column=2)
        Label(conf, text='天\t').grid(row=1, column=3)
        Label(conf, text='\t每天显示时间：').grid(row=2, column=1)
        bdt = Entry(conf, textvariable=self.virus.get_base_day_time_obj(), width=10)
        bdt.grid(row=2, column=2)
        Label(conf, text='秒\t').grid(row=2, column=3)
        Button(conf, text='确定', command=btnOk).grid(row=3, column=1, padx=10, pady=10)
        Button(conf, text='取消', command=conf.destroy).grid(row=3, column=2)

    @staticmethod
    def about():
        """To show author and software information.
        """
        messagebox.showinfo('关于', ABOUT)

    def selectFaFile(self):
        path_ = askopenfilename(filetypes=(("FA files", "*.fa"), ("All files", "*.*")))
        self.fa_path.set(path_)

    def selectCityFile(self):
        path_ = askopenfilename(filetypes=(("Text files", "*.txt"), ("All files", "*.*")))
        self.city_path.set(path_)

    def faster(self):
        self.evolution_speed.set(self.evolution_speed.get() - 1)

    def slower(self):
        self.evolution_speed.set(self.evolution_speed.get() + 1)

    def lastYear(self):
        global INTERVAL
        self.canvas.delete("tags")
        INTERVAL = relativedelta(years=-1)

    def lastMonth(self):
        global INTERVAL
        self.canvas.delete("tags")
        INTERVAL = relativedelta(months=-1)

    @staticmethod
    def pause():
        global RUNNING
        RUNNING = True

    def nextMonth(self):
        global INTERVAL
        self.canvas.delete("tags")
        INTERVAL = relativedelta(months=1)

    def nextYear(self):
        global INTERVAL
        self.canvas.delete("tags")
        INTERVAL = relativedelta(years=1)

    def btnFromNow(self):
        """show from current time
        """
        if self.fa_path.get() == '':
            messagebox.showwarning('提示', '请选择fa文件')
            return
        if self.city_path.get() == '':
            messagebox.showwarning('提示', '请选择城市经纬度文件')
            return
        if '如' in self.from_now.get() or self.from_now == '':
            messagebox.showwarning('提示', '请输入开始时间')
            return

        self.btn_p.config(state="disabled")
        self.btn_fn.config(state="disabled")
        self.btn_s.config(state="disabled")

        oper = Operation(self.fa_path.get(), self.city_path.get())
        index, start_time, end_time = oper.getDetailByTime()
        end_time = self.convertTime(end_time)
        now = self.convertTime(self.from_now.get())

        try:
            self.showVirus(index, now, end_time)

            self.btn_fn.config(state='normal')
            self.btn_p.config(state='normal')
            self.btn_s.config(state="normal")
        except TclError:
            # Tcl Error: After destory the main window,
            # btn_fn, btn_p, btn_s all have benn destoryed,
            # so, there is no config methods.
            exit(0)

    def btnPreprocessing(self):
        """Complete the city file
        """

        if self.fa_path.get() == '':
            messagebox.showwarning('提示', '请选择fa文件')
            return
        if self.city_path.get() == '':
            messagebox.showwarning('提示', '请选择城市经纬度文件')
            return

        oper = Operation(self.fa_path.get(), self.city_path.get())
        flag = oper.judgeIntegrity()
        if flag:
            if len(flag) > UnMatchCityNum:
                with open(LostLocationCities, 'w') as file:
                    file.write(HelpInfo)
                    for city in flag:
                        file.write(' '*29+city+'\n')
                file.close()
                messagebox.showinfo('CityName', '请在'+os.getcwd() +'/'+ LostLocationCities +'中查看待添加经纬度信息的城市')
                
            else:
                for city in flag:
                    self.addCity(self.city_path.get(), city)
        else:
            messagebox.showinfo('Information', '城市经纬度文件完整！')
        

    def btnStart(self):
        """Reconstructing the whole outbreak process
        """
        if self.fa_path.get() == '':
            messagebox.showwarning('提示', '请选择fa文件')
            return
        if self.city_path.get() == '':
            messagebox.showwarning('提示', '请选择城市经纬度文件')
            return

        self.btn_p.config(state="disabled")
        self.btn_fn.config(state="disabled")
        self.btn_s.config(state="disabled")

        oper = Operation(self.fa_path.get(), self.city_path.get())
        index, start_time, end_time = oper.getDetailByTime()
        end_time = self.convertTime(end_time)
        start_time = self.convertTime(start_time)

        try:
            self.showVirus(index, start_time, end_time)

            self.btn_fn.config(state='normal')
            self.btn_p.config(state='normal')
            self.btn_s.config(state="normal")
        except TclError:
            # Tcl Error: After destory the main window,
            # btn_fn, btn_p, btn_s all have benn destoryed,
            # so, there is no config methods.
            exit(0)

    def clearAll(self):
        self.canvas.delete("tags")

    def askQuit(self):
        messagebox.showwarning('Warning', '确认退出 ?')
        self.root.destroy()

    def showClustering(self):
        if self.fa_path.get() == '':
            messagebox.showwarning('提示', '请选择fa文件')
            return

        fc = FrequencyCluster(self.fa_path.get())
        fc.show()


    def showVirus(self, _index, _start_time, _end_time):
        global RUNNING
        global INTERVAL
        total_virus_num = {}
        s_t = time.clock()

        while _start_time <= _end_time + datetime.timedelta(days=self.virus.get_virus_live_time()):
            virus_number = {}
            info = str(_start_time.strftime("%Y/%m/%d"))
            self.canvas.create_text(400, 400, text=info, fill='red', font=('微软雅黑', 18), tags=("tags", "time",))
            death_virus = (_start_time - datetime.timedelta(days=self.virus.get_virus_live_time())).strftime("%Y/%m/%d")
            try:
                today_virus = _index[_start_time.strftime("%Y/%m/%d")]
                for virus in today_virus:
                    y, x = virus[2]
                    x, y = self.convertJingWei(x, y)
                    virus_number[(x, y)] = virus[3]
                    total_virus_num[(x, y)] = []
                    self.canvas.create_text(x, y, text='●', fill="red", tags=("tags", info,))
                    if self.show_city.get():  # show city name
                        self.canvas.create_text(x, y + 10, text=virus[0], fill="red", tags=("tags", info,))
            except KeyError:
                pass

            for k, v in total_virus_num.items():
                if k in virus_number.keys():
                    v.append(virus_number[k])
                else:
                    v.append(0)
            if self.show_virus_num.get():
                for k, v in total_virus_num.items():
                    if sum(v):
                        self.canvas.create_text(k[0] + 5, k[1], text=str(sum(v)), fill="red",
                                                tags=("tags", info, "num"))
            self.canvas.update()

            for key, value in total_virus_num.items():
                if len(value) > self.virus.get_virus_live_time():
                    total_virus_num[key] = value[-self.virus.get_virus_live_time():]

            # for k in list(total_virus_num.keys()):
            #     if sum(total_virus_num[k]) == 0:
            #         del(total_virus_num[k])

            # print('------------------------',info)
            # print(total_virus_num)

            if RUNNING:
                messagebox.showinfo('提示', '已暂停，点击“确定”继续！')
                RUNNING = False

            speed = self.virus.get_base_day_time() / 10 + self.evolution_speed.get() / 10
            if speed < 0:
                messagebox.showwarning('Warning', '已经达到最大速度')
                # If the speed reaches the max, speed < 0, then speed = 0
                self.evolution_speed.set(-self.virus.get_base_day_time())
                speed = 0.0
            time.sleep(speed)  # one day
            self.canvas.delete("num")
            self.canvas.delete(str(death_virus))  # Delete the virus that died on that day
            self.canvas.delete("time")  # Delete the time on map

            _start_time = _start_time + INTERVAL  # wen click the next year,next month button
            INTERVAL = datetime.timedelta(days=1)  # Reset for one day

        total_time = time.clock() - s_t
        messagebox.showinfo('Information', 'Total time: ' + str(total_time) + ' s')

    @staticmethod
    def addCity(_city_file, place):
        """make a window to input and save the Latitude and longitude of city.
        """
        def btnOk():
            if jing.get() == '' or wei.get() == '':
                messagebox.showwarning('Warning', '请输入经纬度信息')
                return
            if 180 < int(jing.get()) or int(jing.get()) < -180:
                messagebox.showerror('Error', '经度范围错误，请输入-180~180之间数字')
                return
            if 90 < int(wei.get()) or int(wei.get()) < -90:
                messagebox.showerror('Error', '纬度范围错误，请输入-90~90之间数字')
                return
            info = ' ' + wei.get() + '     ' + jing.get() + '     ' + place + '\n'
            with open(_city_file, 'a') as cf:
                cf.write(info)
                cf.close()
            messagebox.showinfo('Information', '添加成功')
            window.destroy()

        window = Toplevel()
        window.resizable(False, False)
        window.attributes("-topmost")
        window.title("添加城市经纬度信息")
        Label(window, text='请输入{}的经纬度信息'.format(place)).grid(row=1, column=1)
        Label(window, text=place + ' 的经度：').grid(row=2, column=1)
        jing = Entry(window)
        jing.grid(row=2, column=2, padx=5)
        Label(window, text=place + ' 的纬度：').grid(row=3, column=1)
        wei = Entry(window)
        wei.grid(row=3, column=2, pady=10)
        Button(window, text='添加', command=btnOk).grid(row=4, column=1)
        Button(window, text='取消', command=window.destroy).grid(row=4, column=2)


if __name__ == '__main__':
    app = Application()
    app.addFunc(FunctionMenu)
    app.mainloop()


